server/src/test/java/org/elasticsearch/indices/recovery/IndexRecoveryIT.java

```diff
@@ -863,4 +876,66 @@ public void testHistoryRetention() throws Exception {
         assertThat(recoveryStates.get(0).getIndex().totalFileCount(), is(0));
         assertThat(recoveryStates.get(0).getTranslog().recoveredOperations(), greaterThan(0));
     }
+
+    public void testDoNotInfinitelyWaitForMapping() {
+        internalCluster().startNodes(3, Settings.builder().put("assertion.disabled", RecoveryTarget.class.getName()).build());
+        createIndex("test", Settings.builder()
+            .put("index.analysis.analyzer.test_analyzer.type", "custom")
+            .put("index.analysis.analyzer.test_analyzer.tokenizer", "standard")
+            .putList("index.analysis.analyzer.test_analyzer.filter", "test_token_filter")
+            .put("index.number_of_replicas", 0).put("index.number_of_shards", 1).build());
+        client().admin().indices().preparePutMapping("test")
+            .setType("_doc").setSource("test_field", "type=text,analyzer=test_analyzer").get();
+        int numDocs = between(1, 10);
+        for (int i = 0; i < numDocs; i++) {
+            client().prepareIndex("test", "_doc", "u" + i)
+                .setSource(singletonMap("test_field", Integer.toString(i)), XContentType.JSON).get();
+        }
+        Semaphore recoveryBlocked = new Semaphore(1);
+        for (DiscoveryNode node : clusterService().state().nodes()) {
+            MockTransportService transportService = (MockTransportService) internalCluster().getInstance(
+                TransportService.class, node.getName());
+            transportService.addSendBehavior((connection, requestId, action, request, options) -> {
+                if (action.equals(PeerRecoverySourceService.Actions.START_RECOVERY)) {
+                    if (recoveryBlocked.tryAcquire()) {
+                        PluginsService pluginService = internalCluster().getInstance(PluginsService.class, node.getName());
+                        for (TestAnalysisPlugin plugin : pluginService.filterPlugins(TestAnalysisPlugin.class)) {
```

Comment:
Would it be possible to add a test that hits the regular mapping update required result type, checked for here: https://github.com/elastic/elasticsearch/pull/41099/files#diff-f9ecc51fd8c3001dd782c93d9a040546L350 ?

Reply:
Yes, I will add this test in a follow-up.

Reply:
Thanks!